src/librustc/middle/const_eval.rs

```diff
@@ -706,26 +717,30 @@ pub_fn_checked_op!{ const_uint_checked_shr_via_int(a: u64, b: i64,.. UintTy) {
            uint_shift_body overflowing_shr Uint ShiftRightWithOverflow
 }}
 
-// After type checking, `eval_const_expr_partial` should always suffice. The
-// reason for providing `eval_const_expr_with_substs` is to allow
-// trait-associated consts to be evaluated *during* type checking, when the
-// substs for each expression have not been written into `tcx` yet.
```

Comment:
It might be useful to still include the heart of this comment somewhere.  I.e. having some idea why we need to support const-eval before type-checking is done, and where it arises, is likely to be useful for future people diving into the compiler source code.  The current comment above `EvalHint` doesn't have quite as much info as this comment did in that respect.